research/tensorrt/tensorrt.py

```diff
@@ -0,0 +1,620 @@
+      running an Official Model.
+    output_dir: string representing path to the output directory for saving
+      the frozen graph.
+
+  Returns:
+    Frozen Graph definition for use.
+  """
+  meta_graph = get_serving_meta_graph_def(savedmodel_dir)
+  signature_def = tf.contrib.saved_model.get_signature_def_by_key(
+      meta_graph,
+      tf.saved_model.signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY)
+
+  outputs = [v.name for v in signature_def.outputs.itervalues()]
+  output_names = [node.split(":")[0] for node in outputs]
+
+  graph = tf.Graph()
+  with tf.Session(graph=graph) as sess:
+    tf.saved_model.loader.load(
+        sess, meta_graph.meta_info_def.tags, savedmodel_dir)
+    frozen_graph_def = tf.graph_util.convert_variables_to_constants(
+        sess, graph.as_graph_def(), output_names)
+
+  write_graph_to_file(_GRAPH_FILE, frozen_graph_def, output_dir)
+
+  return frozen_graph_def
+
+
+def get_frozen_graph(graph_file):
+  """Read Frozen Graph file from disk."""
+  with tf.gfile.FastGFile(graph_file, "rb") as f:
+    graph_def = tf.GraphDef()
+    graph_def.ParseFromString(f.read())
+  return graph_def
+
+
+def get_tftrt_name(graph_name, precision_string):
+  return "tftrt_{}_{}".format(precision_string.lower(), graph_name)
+
+
+def get_trt_graph(graph_name, graph_def, precision_mode, output_dir,
+                  output_node, batch_size=128, workspace_size=1<<30):
+  """Create and save inference graph using the TensorRT library.
+
+  Args:
+    graph_name: string, name of the graph to be used for saving.
+    graph_def: GraphDef, the Frozen Graph to be converted.
+    precision_mode: string, the precision that TensorRT should convert into.
+      Options- FP32, FP16, INT8.
+    output_dir: string, the path to where files should be written.
+    output_node: string, the names of the output node that will
+      be returned during inference.
+    batch_size: int, the number of examples that will be predicted at a time.
+    workspace_size: long, size in bytes that can be used during conversion.
+
+  Returns:
+    GraphDef for the TensorRT inference graph.
+  """
+  trt_graph = trt.create_inference_graph(
+      graph_def, [output_node], max_batch_size=batch_size,
+      max_workspace_size_bytes=workspace_size,
+      precision_mode=precision_mode)
+
+  write_graph_to_file(graph_name, trt_graph, output_dir)
+
+  return trt_graph
+
+
+def get_trt_graph_from_calib(graph_name, calib_graph_def, output_dir):
+  """Convert a TensorRT graph used for calibration to an inference graph."""
+  trt_graph = trt.calib_graph_to_infer_graph(calib_graph_def)
+  write_graph_to_file(graph_name, trt_graph, output_dir)
+  return trt_graph
+
+
+################################################################################
+# Run the graph in various precision modes.
+################################################################################
+def get_gpu_config():
+  """Share GPU memory between image preprocessing and inference."""
```

Comment:
I might be wrong, but the image preprocessing and inference sessions don't run at the same time. Is there a reason per_process_gpu_memory_fraction is set to .5?

Reply:
I think that it is because we reserve the memory for each session, and that must not get cleared. I am not sure, but I tried without that part, and got CUDA malloc errors.